### PR TITLE
Fix docbocks ListController behaviour.

### PR DESCRIPTION
**Pull request by jacobdekeizer**
Add correct parameter, return type and exception docblocks for ListController behaviour.

I made `listOverrideColumnValue` return `string|void` like it is in the current codebase. In pr #4972 @bennothommo talked about strict typing for the future, so we could change those return `string|void` to `string|null`. This allows us to add a strict return type `: ?string` in the future. It would not lead to any breaking changes now, but in the future it will when strict return types are added.

Methods that can be changed from `string|void` to `string|null` in this behaviour are:
- `listInjectRowClass`
- `listOverrideColumnValue`
- `listOverrideHeaderValue`

I would like to get your thoughts on this.

### CODE REVIEW

File: modules/backend/behaviors/ListController.php [hunk 1]
```diff
@@ -473,8 +476,9 @@ public function listFilterExtendScopes($host)
 
     /**
      * Controller override: Extend supplied model
-     * @param Model $model
-     * @return Model
+     * @param \October\Rain\Database\Model|\October\Rain\Halcyon\Model $model
+     * @param string|null $definition
+     * @return \October\Rain\Database\Model|\October\Rain\Halcyon\Model
```
Comment (LukeTowers) [inline modules/backend/behaviors/ListController.php]: @jacobdekeizer I don't think that the Lists widget actually supports Halcyon models given the deep integration with the Eloquent query builder

File: modules/backend/behaviors/ListController.php [hunk 2]
```diff
@@ -558,14 +565,15 @@ public static function extendListColumns($callback)
     {
         $calledClass = self::getCalledExtensionClass();
         Event::listen('backend.list.extendColumns', function ($widget) use ($calledClass, $callback) {
+            /** @var \Backend\Widgets\Lists $widget */
```
Comment (LukeTowers) [inline modules/backend/behaviors/ListController.php]: @bennothommo did we decide to allow these inline comments? I'm not sure why they're worth having?

Comment (bennothommo) [reply modules/backend/behaviors/ListController.php]: @LukeTowers I think the intention is to show the expected variable returned from methods which may return several types of classes, like the `makeWidget` method does. It's mainly for providing autocompletion through these types of methods. I don't mind keeping them, but I'd hope their use is very minimal.

Comment (LukeTowers) [reply modules/backend/behaviors/ListController.php]: hmmm, not really a fan. It's pretty clear from the event itself what widget is being returned. @jacobdekeizer thoughts?

Comment (jacobdekeizer) [reply modules/backend/behaviors/ListController.php]: @LukeTowers @bennotjommo It's indeed for autocompletion. It also shows the developer witch widget is returned. We could also change it to `function (\Backend\Widget\Lists $widget) ...`. Is that allowed?

Comment (LukeTowers) [reply modules/backend/behaviors/ListController.php]: @jacobdekeizer hmm, while that would prevent extensibility in terms of being able to inject any other widget in there instead that seems unlikely to be required and I do prefer that in terms of code clarity so I would say go ahead and give it a try and verify that it still works properly.

Comment (jacobdekeizer) [reply modules/backend/behaviors/ListController.php]: @LukeTowers I've changed it and verified that it still works.

File: modules/backend/behaviors/ListController.php [hunk 3]
```diff
@@ -165,7 +165,8 @@ public function makeList($definition = null)
         /*
          * List Widget with extensibility
          */
-        $widget = $this->makeWidget('Backend\Widgets\Lists', $columnConfig);
+        /** @var \Backend\Widgets\Lists $widget */
```
Comment (LukeTowers) [inline modules/backend/behaviors/ListController.php]: Why is this required if the class reference is used to make the widget?

Comment (jacobdekeizer) [reply modules/backend/behaviors/ListController.php]: Because the makeWidget method has as return typehint WidgetBase which doesn't have the methods that are used in the code below it. If we remove this we get many method does not exist errors by a decent IDE.

Comment (LukeTowers) [reply modules/backend/behaviors/ListController.php]: There's a fine line between what the IDE says and common sense by the developer. There's very little reason in the first place for someone to be digging around in this method, and the makeList method already tells the IDE what widget is being returned.

Comment (jacobdekeizer) [reply modules/backend/behaviors/ListController.php]: In my opinion good readable codebase should be the goal of every project. This includes the readability by the ide. Why would you want a file with 10 lines of errors by the ide or analyzer tool like phpstan. Phpstan also reads these kind of docblocks if I recall correctly. Coming from laravel I find it somewhat shocking to look into the octoberms codebase. So I'm trying to help you guys fixing some things up by making the code readable for an ide and a person.

Comment (LukeTowers) [reply modules/backend/behaviors/ListController.php]: I'm going to defer to @daftspunk's opinion on this one, I personally don't like having `@var` comments littered all over the codebase since they provide little benefit in my eyes. I appreciate your effort in cleaning things up a bit though and your patience as we decide how we want to handle these sort of additions.

Comment (Klaasie) [reply modules/backend/behaviors/ListController.php]: >  There's very little reason in the first place for someone to be digging around in this method

My personal opinion:
I find this odd. We should be encouraging people to get to know the system to such a degree that they end up digging around this method (any method like that really). It means they're looking to understand OC as a system. For a project that (I think) would like to have more contributions from the community it'd be great if they understand most of the system.

Having that said, allowing these kind of docblock lines will help most of the major IDE's understand exactly what the code is doing and enable functionality like jump to definition. It would help people understand the system at a faster pace.

list widget setSearchTerm

As an example, this is what my IDE currently knows about the method `setSearchTerm` (which is pretty much just listing any class having the method). While we could tell the IDE it's the first one.

Comment (daftspunk) [reply modules/backend/behaviors/ListController.php]: Is there no way to train the makeWidget method to have a return type of the first argument?

Comment (jacobdekeizer) [reply modules/backend/behaviors/ListController.php]: @daftspunk nog that I know of, since we don't have generics in PHP.

Comment (daftspunk) [reply modules/backend/behaviors/ListController.php]: If you change it from returning WidgetBase to a generic object, will it stop the errors?

The proposed solution is a WET monkey patch

Comment (jacobdekeizer) [reply modules/backend/behaviors/ListController.php]: @daftspunk No that wouldn't fix the errors. It's good that makeWidget returns WidgetBase, so I wouldn't change that.

Comment (daftspunk) [reply modules/backend/behaviors/ListController.php]: It's not returning a WidgetBase. It's returning a generic object that inherits WidgetBase. Unfortunately the IDE is not smart enough to figure this out. I wonder how it would handle magic methods?

The issue is not everyone uses an IDE and this line of code makes no sense, and cannot be verified to be correct, so we will constantly face this problem of "incorrectness". You must treat the problem and not the symptom. Someh

Comment (jacobdekeizer) [reply modules/backend/behaviors/ListController.php]: @daftspunk Phpstorm isn't a fan of magic method either 😄 . I don't know how you want me to fix it. If we remove the docblock my IDE warns about this, and not the accessed methods I talked about earlier (sorry for that).
afbeelding
Changing the return type of this method to WidgetBase also feels wrong, then I'm leaning more towards leaving this docblock out and accept the warning of Phpstorm.

Comment (daftspunk) [reply modules/backend/behaviors/ListController.php]: I understand. The IDE helpers can be  overly strict which can stifle innovation; the things that make PHP special and October possible in the first place

My preference would be to make the return value generic and sacrifice the intellisense, so we don't keep revisiting this issue. But, since you are leading this initiative as an IDE user, we can follow your preference

It is a shame but hopefully the IDE will improve to support more flexibility in dynamic languages, or maybe we will think of a workable solution in the meantime, one that addresses the problem at its source

Comment (jacobdekeizer) [reply modules/backend/behaviors/ListController.php]: @daftspunk what do you think about changing string|void to string|null as I talked about in the pull request comment.

Comment (jacobdekeizer) [reply modules/backend/behaviors/ListController.php]: @daftspunk another possibility is to change the return type of `\Backend\Widgets\Lists` to `\Backend\Classes\WidgetBase|\Backend\Widgets\Lists`. Then we havo no errors in the code

Comment (LukeTowers) [reply modules/backend/behaviors/ListController.php]: @jacobdekeizer that's worse, because now we have to start listing every single possible implementation of WidgetBase.

Comment (jacobdekeizer) [reply modules/backend/behaviors/ListController.php]: @LukeTowers @daftspunk I'm still against removing @var tbh. Why can't we help the ide out so we have good autocompletion. I saw in pr #4960 someone also used @var notation to help the ide out. I don't think it's bad practice, it's the nature of php that we have to do this.

Comment (daftspunk) [reply modules/backend/behaviors/ListController.php]: As discussed, the method returns `object|\Backend\Classes\WidgetBase`. We want to resolve IDE errors as first priority, autocomplete is secondary

Comment (jacobdekeizer) [reply modules/backend/behaviors/ListController.php]: The @var notation fixes both issues. We fix the ide errors and it returns the correct type.

Comment (daftspunk) [reply modules/backend/behaviors/ListController.php]: > The issue is not everyone uses an IDE and this line of code makes no sense, and cannot be verified to be correct, so we will constantly face this problem of "incorrectness". You must treat the cause and not the symptom

Comment (jacobdekeizer) [reply modules/backend/behaviors/ListController.php]: @daftspunk if someone doesn't use an ide, then all the code is dus plain text. What difference does that one line then make? Instead all the ide users get good typehinting and autocompletion. We can return WidgetBase, but in my opinion adding `@var` and returning the  Lists object would be so much better and the only right thing to do. But if you guys are so against that one line then I will remove it.

Comment (daftspunk) [reply modules/backend/behaviors/ListController.php]: Exactly, it is an opinion that says "Use PHPStorm/IDE". It's not just one line. It's a modification of existing architecture where that line must be introduced everywhere. Even in your own Pull Request, it is used some places and not others, how to verify this? How to maintain this?

Thank you for your understanding

**Comment by daftspunk** [comment]
Thanks @jacobdekeizer!
